Fix the scenario tests failure

branch:
stable/mitaka

**Description**
All the scenario tests fail with this error message:
Traceback (most recent call last):
  File "/Users/zakeri/GitHub/neutron-lbaas/neutron_lbaas/tests/tempest/v2/scenario/base.py", line 85, in setUp
    self._set_net_and_subnet()
  File "/Users/zakeri/GitHub/neutron-lbaas/neutron_lbaas/tests/tempest/v2/scenario/base.py", line 134, in _set_net_and_subnet
    config.compute.fixed_network_name)
  File "/Users/zakeri/GitHub/neutron-lbaas/.tox/apiv2/lib/python2.7/site-packages/tempest/scenario/manager.py", line 857, in _get_network_by_name
    "Unable to get network by name: %s" % network_name)
  File "/Users/zakeri/GitHub/neutron-lbaas/.tox/apiv2/lib/python2.7/site-packages/unittest2/case.py", line 845, in assertNotEqual
    raise self.failureException(msg)
_pytest.assertion.reinterpret.AssertionError: 0 == 0 : Unable to get network by name: None

**Tests:**

neutron_lbaas/tests/tempest/v2/scenario/test_healthmonitor_basic.py::TestHealthMonitorBasic::test_health_monitor_basic <- .tox/apiv2/lib/python2.7/site-packages/tempest/test.py FAILED
neutron_lbaas/tests/tempest/v2/scenario/test_listener_basic.py::TestListenerBasic::test_listener_basic <- .tox/apiv2/lib/python2.7/site-packages/tempest/test.py FAILED
neutron_lbaas/tests/tempest/v2/scenario/test_load_balancer_basic.py::TestLoadBalancerBasic::test_load_balancer_basic <- .tox/apiv2/lib/python2.7/site-packages/tempest/test.py FAILED
neutron_lbaas/tests/tempest/v2/scenario/test_load_balancer_tls.py::TestLoadBalancerTLS::test_load_balancer_tls <- .tox/apiv2/lib/python2.7/site-packages/tempest/test.py FAILED
neutron_lbaas/tests/tempest/v2/scenario/test_session_persistence.py::TestSessionPersistence::test_session_persistence <- .tox/apiv2/lib/python2.7/site-packages/tempest/test.py FAILED
neutron_lbaas/tests/tempest/v2/scenario/test_shared_pools.py::TestSharedPools::test_shared_pools <- .tox/apiv2/lib/python2.7/site-packages/tempest/test.py FAILED

--- a/neutron_lbaas/tests/tempest/v2/api/base.py
+++ b/neutron_lbaas/tests/tempest/v2/api/base.py
@@ -18,6 +18,7 @@ import time
 from oslo_log import log as logging
 from tempest.api.network import base
 from tempest import config
+from tempest.lib.common.utils import test_utils
 from tempest.lib import exceptions
 
 from neutron_lbaas._i18n import _, _LI
@@ -26,6 +27,7 @@ from neutron_lbaas.tests.tempest.v2.clients import listeners_client
 from neutron_lbaas.tests.tempest.v2.clients import load_balancers_client
 from neutron_lbaas.tests.tempest.v2.clients import members_client
 from neutron_lbaas.tests.tempest.v2.clients import pools_client
+import time
 
 CONF = config.CONF
 
@@ -90,23 +92,33 @@ class BaseTestCase(base.BaseNetworkTest):
                 for pool in listener.get('pools'):
                     hm = pool.get('healthmonitor')
                     if hm:
-                        cls._try_delete_resource(
+                        test_utils.call_and_ignore_notfound_exc(
                             cls.health_monitors_client.delete_health_monitor,
                             pool.get('healthmonitor').get('id'))
                         cls._wait_for_load_balancer_status(lb_id)
-                    cls._try_delete_resource(cls.pools_client.delete_pool,
-                                             pool.get('id'))
+                    test_utils.call_and_ignore_notfound_exc(
+                        cls.pools_client.delete_pool,
+                        pool.get('id'))
                     cls._wait_for_load_balancer_status(lb_id)
-                    health_monitor = pool.get('healthmonitor')
-                    if health_monitor:
-                        cls._try_delete_resource(
-                            cls.health_monitors_client.delete_health_monitor,
-                            health_monitor.get('id'))
+                    # delete pool's members
+                    members = pool.get('members', [])
+                    for member in members:
+                        test_utils.call_and_ignore_notfound_exc(
+                            cls.members_client.delete_member,
+                            pool.get('id'), member.get('id'))
+                        cls._wait_for_load_balancer_status(lb_id)
+                    # delete pool
+                    test_utils.call_and_ignore_notfound_exc(
+                        cls.pools_client.delete_pool, pool.get('id'))
                     cls._wait_for_load_balancer_status(lb_id)
-                cls._try_delete_resource(cls.listeners_client.delete_listener,
-                                         listener.get('id'))
+                # delete listener
+                test_utils.call_and_ignore_notfound_exc(
+                    cls.listeners_client.delete_listener,
+                    listener.get('id'))
                 cls._wait_for_load_balancer_status(lb_id)
-            cls._try_delete_resource(cls._delete_load_balancer, lb_id)
+            # delete load-balancer
+            test_utils.call_and_ignore_notfound_exc(
+                cls._delete_load_balancer, lb_id)
 
         super(BaseTestCase, cls).resource_cleanup()
 

--- a/neutron_lbaas/tests/tempest/v2/api/test_listeners_tls.py
+++ b/neutron_lbaas/tests/tempest/v2/api/test_listeners_tls.py
@@ -23,7 +23,6 @@ from tempest.lib.common.utils import data_utils
 from tempest.lib import decorators
 from tempest.lib import exceptions as ex
 from tempest import test
-
 from neutron_lbaas.tests.tempest.v2.api import base
 from neutron_lbaas.tests.tempest.v2.common import barbican_helper
 

--- a/neutron_lbaas/tests/tempest/v2/api/test_listeners_tls_sni.py
+++ b/neutron_lbaas/tests/tempest/v2/api/test_listeners_tls_sni.py
@@ -1,0 +1,141 @@
+# Copyright 2016 Rackspace US Inc.
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from barbicanclient import client as barbican_client
+from keystoneauth1 import session
+from keystoneauth1 import identity
+from keystoneclient.v2_0 import client as v2_client
+from keystoneclient.v3 import client as v3_client
+from tempest import config
+from tempest.lib.common.utils import data_utils
+from tempest.lib import decorators
+from tempest.lib import exceptions as ex
+from tempest import test
+from neutron_lbaas.tests.tempest.v2.api import base
+from neutron_lbaas.tests.tempest.v2.common import barbican_helper
+
+config = config.CONF
+
+
+class TLSSNIListenersTestJSON(base.BaseTestCase):
+
+    """
+    Tests the following operations in the Neutron-LBaaS API using the
+    REST client for Listeners with TLS:
+
+        list listeners
+        create listener
+        get listener
+        update listener
+        delete listener
+    """
+
+    @classmethod
+    def resource_setup(cls):
+        super(TLSSNIListenersTestJSON, cls).resource_setup()
+        if not test.is_extension_enabled('lbaas', 'network'):
+            msg = "lbaas extension not enabled."
+            raise cls.skipException(msg)
+        network_name = data_utils.rand_name('network-')
+        cls.network = cls.create_network(network_name)
+        cls.subnet = cls.create_subnet(cls.network)
+        cls.create_lb_kwargs = {'tenant_id': cls.subnet['tenant_id'],
+                                'vip_subnet_id': cls.subnet['id']}
+        cls.load_balancer = cls._create_active_load_balancer(
+            **cls.create_lb_kwargs)
+        cls.protocol = 'TERMINATED_HTTPS'
+        cls.port = 443
+        cls.load_balancer_id = cls.load_balancer['id']
+
+        # Barbican setup
+        client_kwargs = {
+            'username': config.auth.admin_username,
+            'password': config.auth.admin_password,
+            'tenant_name': config.auth.admin_project_name,
+            'auth_url': config.identity.uri
+        }
+        if config.identity.auth_version.lower() == 'v2':
+            cls.keystone_client = v2_client.Client(**client_kwargs)
+        elif config.identity.auth_version.lower() == 'v3':
+            client_kwargs['auth_url'] = config.identity.uri_v3
+            cls.keystone_client = v3_client.Client(**client_kwargs)
+        else:
+            raise Exception('Unknown authentication version')
+
+        # TODO(): Keystone should load service endpoint...
+        cls.auth = identity.v2.Password(**client_kwargs)
+
+        # create a Keystone session using the auth plugin we just created
+        keystone_session = session.Session(auth=cls.auth)
+
+        cls.barbican_client = barbican_client.Client(session=keystone_session)
+
+        cls.barbican_helper = barbican_helper.BarbicanTestHelper(
+            cls.barbican_client)
+        tls_container_ref = cls.barbican_helper.create_barbican_resources()
+        second_tls_container_ref = cls.barbican_helper.create_barbican_resources()
+        cls.default_tls_container_ref = tls_container_ref
+        cls.sni_tls_container_ref = second_tls_container_ref
+
+        cls.create_listener_kwargs = {
+            'loadbalancer_id': cls.load_balancer_id,
+            'protocol': cls.protocol,
+            'protocol_port': cls.port,
+            'default_tls_container_ref': cls.default_tls_container_ref,
+            'sni_container_refs': cls.sni_tls_container_ref}
+        cls.listener = cls._create_listener(**cls.create_listener_kwargs)
+        cls.listener_id = cls.listener['id']
+
+
+
+    @test.attr(type='smoke')
+    def test_create_tls_sni_listener(self):
+        """Test create tls listener"""
+        create_new_listener_kwargs = self.create_listener_kwargs
+        create_new_listener_kwargs['protocol_port'] = 8081
+        new_listener = self._create_listener(
+            **create_new_listener_kwargs)
+        new_listener_id = new_listener['id']
+        self.addCleanup(self._delete_listener, new_listener_id)
+        self._check_status_tree(
+            load_balancer_id=self.load_balancer_id,
+            listener_ids=[self.listener_id, new_listener_id])
+        listener = self.listeners_client.get_listener(
+            new_listener_id)
+        self.assertEqual(new_listener, listener)
+        self.assertNotEqual(self.listener, new_listener)
+
+
+    @test.attr(type='smoke')
+    def test_delete_tls_sni_listener(self):
+        """Test delete listener"""
+        new_tls_container_ref = (
+            self.barbican_helper.create_barbican_resources())
+        create_new_listener_kwargs = self.create_listener_kwargs
+        create_new_listener_kwargs['default_tls_container_ref'] = (
+            new_tls_container_ref)
+        new_listener = self._create_listener(**create_new_listener_kwargs)
+        new_listener_id = new_listener['id']
+        self._check_status_tree(
+            load_balancer_id=self.load_balancer_id,
+            listener_ids=[self.listener_id, new_listener_id])
+        listener = self.listeners_client.get_listener(
+            new_listener_id)
+        self.assertEqual(new_listener, listener)
+        self.assertNotEqual(self.listener, new_listener)
+        self._delete_listener(new_listener_id)
+        self.assertRaises(ex.NotFound,
+                          self.listeners_client.get_listener,
+                          new_listener_id)

--- a/neutron_lbaas/tests/tempest/v2/scenario/base.py
+++ b/neutron_lbaas/tests/tempest/v2/scenario/base.py
@@ -13,14 +13,14 @@
 #    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 #    License for the specific language governing permissions and limitations
 #    under the License.
-
+import ssl
 import select
 import shlex
 import socket
 import subprocess
 import tempfile
 import time
-
+from tempest.lib.common import ssh
 from oslo_log import log as logging
 import six
 from six.moves import http_cookiejar
@@ -29,11 +29,11 @@ from six.moves.urllib import request as urllib2
 from tempest.common import waiters
 from tempest import config
 from tempest import exceptions
-from tempest.lib.common import ssh
 from tempest.lib import exceptions as lib_exc
 from tempest.scenario import manager
-
 from tempest import test
+from tempest.lib.common.utils import test_utils
+import requests
 
 from neutron_lbaas._i18n import _
 from neutron_lbaas.tests.tempest.v2.clients import health_monitors_client
@@ -72,7 +72,6 @@ class BaseTestCase(manager.NetworkScenarioTest):
 
     def setUp(self):
         super(BaseTestCase, self).setUp()
-
         self.servers_keypairs = {}
         self.servers = {}
         self.members = []
@@ -110,7 +109,7 @@ class BaseTestCase(manager.NetworkScenarioTest):
             msg = 'LBaaS Extension is not enabled'
             raise cls.skipException(msg)
         if not (cfg.project_networks_reachable or cfg.public_network_id):
-            msg = ('Either project_networks_reachable must be "true", or '
+            msg = ('Either tenant_networks_reachable must be "true", or '
                    'public_network_id must be defined.')
             raise cls.skipException(msg)
 
@@ -127,8 +126,10 @@ class BaseTestCase(manager.NetworkScenarioTest):
             tenant_net = None
 
         if tenant_net:
-            tenant_subnet = self._list_subnets(tenant_id=self.tenant_id)[0]
-            self.subnet = tenant_subnet['subnet']
+            self.subnet = self._list_subnets(tenant_id=self.tenant_id)[0]
+            self.addCleanup(test_utils.call_and_ignore_notfound_exc,
+                            self.networks_client.delete_network,
+                            self.subnet['id'])
             self.network = tenant_net
         else:
             self.network = self._get_network_by_name(
@@ -137,7 +138,8 @@ class BaseTestCase(manager.NetworkScenarioTest):
             # with the fixed network is the one we want.  In the future, we
             # should instead pull a subnet id from config, which is set by
             # devstack/admin/etc.
-            self.subnet = self._list_subnets(network_id=self.network['id'])[0]
+            subnet = self._list_subnets(network_id=self.network['id'])[0]
+            self.subnet = subnet
 
     def _create_security_group_for_test(self):
         self.security_group = self._create_security_group(
@@ -192,8 +194,8 @@ class BaseTestCase(manager.NetworkScenarioTest):
             floating_ip = self.create_floating_ip(
                 server, public_network_id)
             self.check_floating_ip_status(floating_ip, "ACTIVE")
-            self.floating_ips[floating_ip] = server
-            self.server_ips[server['id']] = floating_ip.floating_ip_address
+            self.floating_ips[floating_ip['id']] = server
+            self.server_ips[server['id']] = floating_ip['floating_ip_address']
         else:
             self.server_ips[server['id']] =\
                 server['addresses'][net_name][0]['addr']
@@ -233,17 +235,13 @@ class BaseTestCase(manager.NetworkScenarioTest):
         """
         LOG.info(('self.server_ips looks like this: {0}'.format(
             self.server_ips)))
-
         for server_id, ip in six.iteritems(self.server_ips):
-
             LOG.info(('processing server_id {0}  ip: {1}'.format(
                 server_id, ip)))
-
             private_key = self.servers_keypairs[server_id]['private_key']
             server = self.servers_client.show_server(server_id)['server']
             server_name = server['name']
             username = config.validation.image_ssh_user
-
             self.ssh_client = ssh.Client(host=ip, username=username,
                                          pkey=private_key)
             self.ssh_client.test_connection_auth()
@@ -265,6 +263,7 @@ class BaseTestCase(manager.NetworkScenarioTest):
                                            "/tmp/script1",
                                            ip,
                                            username, key.name)
+
             # Start netcat
             start_server = ('while true; do '
                             'sudo nc -ll -p %(port)s -e sh /tmp/%(script)s; '
@@ -272,8 +271,6 @@ class BaseTestCase(manager.NetworkScenarioTest):
             cmd = start_server % {'port': self.port1,
                                   'script': 'script1'}
             self.exec_command(cmd)
-            LOG.info(('Executing cmd on server {0} ip {1}: {2}'.format(
-                server_id, ip, cmd)))
 
             if len(self.server_ips) == 1:
                 with tempfile.NamedTemporaryFile() as script:
@@ -332,23 +329,27 @@ class BaseTestCase(manager.NetworkScenarioTest):
         return self.pool
 
     def _cleanup_load_balancer(self, load_balancer_id):
-        self.delete_wrapper(self.load_balancers_client.delete_load_balancer,
-                            load_balancer_id)
+        test_utils.call_and_ignore_notfound_exc(
+            self.load_balancers_client.delete_load_balancer, load_balancer_id)
+#        self.delete_wrapper(self.load_balancers_client.delete_load_balancer,
+#                            load_balancer_id)
         self._wait_for_load_balancer_status(load_balancer_id, delete=True)
 
     def _cleanup_listener(self, listener_id, load_balancer_id=None):
-        self.delete_wrapper(self.listeners_client.delete_listener, listener_id)
+        test_utils.call_and_ignore_notfound_exc(
+            self.listeners_client.delete_listener, listener_id)
         if load_balancer_id:
             self._wait_for_load_balancer_status(load_balancer_id)
 
     def _cleanup_pool(self, pool_id, load_balancer_id=None):
-        self.delete_wrapper(self.pools_client.delete_pool, pool_id)
+        test_utils.call_and_ignore_notfound_exc(
+            self.pools_client.delete_pool, pool_id)
         if load_balancer_id:
             self._wait_for_load_balancer_status(load_balancer_id)
 
     def _cleanup_health_monitor(self, hm_id, load_balancer_id=None):
-        self.delete_wrapper(self.health_monitors_client.delete_health_monitor,
-                            hm_id)
+        test_utils.call_and_ignore_notfound_exc(
+            self.health_monitors_client.delete_health_monitor, hm_id)
         if load_balancer_id:
             self._wait_for_load_balancer_status(load_balancer_id)
 
@@ -386,11 +387,11 @@ class BaseTestCase(manager.NetworkScenarioTest):
 
     def _assign_floating_ip_to_lb_vip(self, lb):
         public_network_id = config.network.public_network_id
-        port_id = lb.vip_port_id
+        port_id = lb['vip_port_id']
         floating_ip = self.create_floating_ip(lb, public_network_id,
                                               port_id=port_id)
-        self.floating_ips.setdefault(lb.id, [])
-        self.floating_ips[lb.id].append(floating_ip)
+        self.floating_ips.setdefault(lb['id'], [])
+        self.floating_ips[lb['id']].append(floating_ip)
         # Check for floating ip status before you check load-balancer
         self.check_floating_ip_status(floating_ip, "ACTIVE")
 
@@ -429,10 +430,10 @@ class BaseTestCase(manager.NetworkScenarioTest):
         # tempest.conf file
         if ip_version == 4:
             if (config.network.public_network_id and not
-                    config.network.project_networks_reachable):
+                        config.network.project_networks_reachable):
                 self._assign_floating_ip_to_lb_vip(self.load_balancer)
                 self.vip_ip = self.floating_ips[
-                    self.load_balancer.id][0]['floating_ip_address']
+                    self.load_balancer['id']][0]['floating_ip_address']
 
         # Currently the ovs-agent is not enforcing security groups on the
         # vip port - see https://bugs.launchpad.net/neutron/+bug/1163569
@@ -440,7 +441,7 @@ class BaseTestCase(manager.NetworkScenarioTest):
         # security group with a rule that allows tcp port 80 to the vip port.
         self.ports_client.update_port(
             self.load_balancer.get('vip_port_id'),
-            security_groups=[self.security_group.id])
+            security_groups=[self.security_group['id']])
 
     def _wait_for_load_balancer_status(self, load_balancer_id,
                                        provisioning_status='ACTIVE',
@@ -505,11 +506,11 @@ class BaseTestCase(manager.NetworkScenarioTest):
                   type=sp_type))
 
     def _check_load_balancing(self, protocol='http', port=80):
-        """Check Load Balancing between 2 servers
-
+        """
         1. Send NUM requests on the floating ip associated with the VIP
         2. Check that the requests are shared between the two servers
         """
+
         self._check_connection(check_ip=self.vip_ip,
                                protocol=protocol,
                                port=port)
@@ -520,24 +521,36 @@ class BaseTestCase(manager.NetworkScenarioTest):
             self.assertGreater(counter, 0, 'Member %s never balanced' % member)
 
     def _check_connection(self, check_ip, protocol='http', port=80):
-        """Checks connection of the back end servers"""
+        #ssl.PROTOCOL_SSLv23 = ssl.PROTOCOL_TLSv1
+        #context = ssl._create_unverified_context
+        #default_context = ssl._create_default_https_context
+
         def try_connect(check_ip, protocol, port):
+            requests.packages.urllib3.disable_warnings()
+            #print check_ip
+            #url = "{0}://{1}:{2}/".format(protocol, check_ip, port)
             try:
+                #resp = requests.get("{0}://{1}:{2}/".format(protocol,
+                #                                            check_ip,
+                #                                            port), auth=('cirros','cubswin:)'))
+
                 resp = urllib2.urlopen("{0}://{1}:{2}/".format(protocol,
-                                                               check_ip,
-                                                               port))
-                if resp.getcode() == 200:
+                                                            check_ip,
+                                                            port))
+                print resp.status_code
+                if resp.status_code == 200 :
                     return True
                 return False
             except IOError:
                 return False
             except error.HTTPError:
                 return False
+
         timeout = config.validation.ping_timeout
         start = time.time()
         while not try_connect(check_ip, protocol, port):
             if (time.time() - start) > timeout:
-                message = "Timed out trying to connect to %s" % check_ip
+                message = "Timed out trying to connect to %s " %check_ip
                 raise exceptions.TimeoutException(message)
 
     def _send_requests(self, vip_ip, servers, protocol='http'):
@@ -565,7 +578,6 @@ class BaseTestCase(manager.NetworkScenarioTest):
 
     def _check_load_balancing_after_deleting_resources(self):
         """Check that the requests are not sent to any servers
-
         Assert that no traffic is sent to any servers
         """
         counters = self._send_requests(self.vip_ip, ["server1", "server2"])
@@ -716,4 +728,3 @@ class BaseTestCase(manager.NetworkScenarioTest):
                     command=cmd, exit_status=exit_status,
                     stderr=err_data, stdout=out_data)
             return out_data
-

--- a/neutron_lbaas/tests/tempest/v2/scenario/base_tls.py
+++ b/neutron_lbaas/tests/tempest/v2/scenario/base_tls.py
@@ -20,7 +20,7 @@ from keystoneclient.v2_0 import client as v2_client
 from keystoneclient.v3 import client as v3_client
 from oslo_log import log as logging
 from tempest import config
-
+from keystoneauth1 import identity
 from neutron_lbaas.tests.tempest.v2.common import barbican_helper
 from neutron_lbaas.tests.tempest.v2.scenario import base
 
@@ -36,7 +36,7 @@ class BaseTestCaseTLS(base.BaseTestCase):
         client_kwargs = {
             'username': config.auth.admin_username,
             'password': config.auth.admin_password,
-            'tenant_name': config.auth.admin_tenant_name,
+            'tenant_name': config.auth.admin_project_name,
             'auth_url': config.identity.uri
         }
         if config.identity.auth_version.lower() == 'v2':
@@ -49,9 +49,12 @@ class BaseTestCaseTLS(base.BaseTestCase):
         keystone_session = session.Session(auth=self.keystone_client)
 
         # TODO(): Keystone should load service endpoint...
-        self.temp_barbican_endpoint = 'http://localhost:9311'
-        self.barbican_client = barbican_client.Client(
-            session=keystone_session, endpoint=self.temp_barbican_endpoint)
+        self.auth = identity.v2.Password(**client_kwargs)
+
+        # create a Keystone session using the auth plugin we just created
+        keystone_session = session.Session(auth=self.auth)
+
+        self.barbican_client = barbican_client.Client(session=keystone_session)
 
         self.barbican_helper = barbican_helper.BarbicanTestHelper(
             self.barbican_client)

--- a/neutron_lbaas/tests/tempest/v2/scenario/test_loadbalancer_tls.py
+++ b/neutron_lbaas/tests/tempest/v2/scenario/test_loadbalancer_tls.py
@@ -1,0 +1,52 @@
+# Copyright 2015, 2016 Rackspace Inc.
+# Copyright 2016 Hewlett Packard Enterprise Development Company
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+
+from tempest import config
+from tempest import test
+
+from neutron_lbaas.tests.tempest.v2.scenario import base_tls
+
+config = config.CONF
+
+
+class TestLoadBalancerTLS(base_tls.BaseTestCaseTLS):
+
+    """This test checks TLS load balancing
+
+    The following is the scenario outline:
+    1. Create an instance
+    2. SSH to the instance and start two servers
+    3. Create a TLS termination load balancer with two members and
+       ROUND_ROBIN algorithm, associate the VIP with a floating ip
+    4. Send NUM requests to the floating ip and check that they are shared
+       between the two servers.
+    5. SNI (Create additional certificate(s), add to listener sni, verify
+       returned certificates)
+    """
+    @test.services('compute', 'network')
+    def test_load_balancer_tls(self):
+        # Test TLS termination
+        # Verify HaProxy terminates TLS traffic and sends it to the backend
+        self._create_server('server1')
+        self._start_servers()
+        default_tls_container_ref = self._create_barbican_resources()
+        self._create_load_balancer(
+            default_tls_container_ref = default_tls_container_ref,
+            listener_protocol='TERMINATED_HTTPS',
+            port=443,
+            pool_protocol='HTTP')
+        self._check_load_balancing(protocol='https', port=443)
+

